File: sample-tests/kotlin/src/test/kotlin/com/icure/docs/generator/MdxExtractorTest.kt
package com.icure.docs.generator

import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain

class MdxExtractorTest : FunSpec({

    test("extracts a single Kotlin block") {
        val mdx = """
            Some text
            ```kotlin
            val x = 1
            ```
            More text
        """.trimIndent()

        val blocks = extractKotlinBlocks(mdx)
        blocks shouldHaveSize 1
        blocks[0].code shouldBe "val x = 1"
        blocks[0].startLine shouldBe 2
        blocks[0].testId shouldBe null
    }

    test("extracts multiple Kotlin blocks") {
        val mdx = """
            ```kotlin
            val a = 1
            ```
            text
            ```kotlin
            val b = 2
            ```
        """.trimIndent()

        val blocks = extractKotlinBlocks(mdx)
        blocks shouldHaveSize 2
    }

    test("skips no-test blocks") {
        val mdx = """
            ```kotlin no-test
            val skip = true
            ```
            ```kotlin
            val keep = true
            ```
        """.trimIndent()

        val blocks = extractKotlinBlocks(mdx)
        blocks shouldHaveSize 1
        blocks[0].code shouldContain "keep"
    }

    test("extracts test ID from fence info") {
        val mdx = """
            ```kotlin test-ABCD
            val x = 1
            ```
        """.trimIndent()

        val blocks = extractKotlinBlocks(mdx)
        blocks shouldHaveSize 1
        blocks[0].testId shouldBe "ABCD"
    }

    test("ignores non-Kotlin blocks") {
        val mdx = """
            ```typescript
            const x = 1
            ```
            ```python
            x = 1
            ```
            ```kotlin
            val x = 1
            ```
        """.trimIndent()

        val blocks = extractKotlinBlocks(mdx)
        blocks shouldHaveSize 1
    }

    test("returns empty for no Kotlin blocks") {
        val mdx = """
            ```typescript
            const x = 1
            ```
        """.trimIndent()

        val blocks = extractKotlinBlocks(mdx)
        blocks.shouldBeEmpty()
    }

    test("extracts multi-line block") {
        val mdx = """
            ```kotlin
            import com.icure.cardinal.sdk.CardinalSdk

            suspend fun foo(sdk: CardinalSdk): String {
                return "hello"
            }
            ```
        """.trimIndent()

        val blocks = extractKotlinBlocks(mdx)
        blocks shouldHaveSize 1
        blocks[0].code shouldContain "suspend fun foo"
        blocks[0].code shouldContain "import com.icure"
    }

    test("isKotlinImportLine detects imports") {
        isKotlinImportLine("import com.icure.cardinal.sdk.CardinalSdk") shouldBe true
        isKotlinImportLine("  import com.icure.cardinal.sdk.CardinalSdk") shouldBe true
        isKotlinImportLine("// import com.icure.cardinal.sdk.CardinalSdk") shouldBe false
        isKotlinImportLine("val x = 1") shouldBe false
    }

    test("stripImportsFromBlock removes imports and trims") {
        val code = """
            import com.icure.cardinal.sdk.CardinalSdk
            import com.icure.cardinal.sdk.model.DecryptedPatient

            suspend fun foo(sdk: CardinalSdk) {
                val p = DecryptedPatient()
            }
        """.trimIndent()

        val stripped = stripImportsFromBlock(code)
        stripped shouldBe """
            suspend fun foo(sdk: CardinalSdk) {
                val p = DecryptedPatient()
            }
        """.trimIndent()
    }

    test("stripImportsFromBlock strips private and const modifiers") {
        val code = """
            private const val MY_URL = "https://api.icure.cloud"
            private val secret = "abc"
            public fun doSomething() {}
        """.trimIndent()

        val stripped = stripImportsFromBlock(code)
        stripped shouldBe """
            val MY_URL = "https://api.icure.cloud"
            val secret = "abc"
            fun doSomething() {}
        """.trimIndent()
    }
})

File: sample-tests/typescript/src/__tests__/mdx-extractor.test.ts
import { describe, test, expect } from 'vitest'
import { extractTypeScriptBlocks, isImportLine, stripImportsFromBlock } from '../mdx-extractor'

describe('extractTypeScriptBlocks', () => {
  test('extracts a single typescript block with line number', () => {
    const mdx = [
      '# Title',
      '',
      '```typescript',
      'const x = 1',
      '```',
    ].join('\n')

    const blocks = extractTypeScriptBlocks(mdx)
    expect(blocks).toHaveLength(1)
    expect(blocks[0].code).toBe('const x = 1')
    expect(blocks[0].startLine).toBe(3)
  })

  test('extracts multiple blocks preserving order', () => {
    const mdx = [
      '```typescript',
      'const a = 1',
      '```',
      '',
      'Some text',
      '',
      '```typescript',
      'const b = 2',
      '```',
    ].join('\n')

    const blocks = extractTypeScriptBlocks(mdx)
    expect(blocks).toHaveLength(2)
    expect(blocks[0].code).toBe('const a = 1')
    expect(blocks[1].code).toBe('const b = 2')
  })

  test('skips blocks with no-test annotation', () => {
    const mdx = [
      '```typescript no-test',
      'const x = 1',
      '```',
      '',
      '```typescript',
      'const y = 2',
      '```',
    ].join('\n')

    const blocks = extractTypeScriptBlocks(mdx)
    expect(blocks).toHaveLength(1)
    expect(blocks[0].code).toBe('const y = 2')
  })

  test('ignores non-typescript fenced blocks', () => {
    const mdx = [
      '```kotlin',
      'val x = 1',
      '```',
      '',
      '```python',
      'x = 1',
      '```',
      '',
      '```typescript',
      'const x = 1',
      '```',
    ].join('\n')

    const blocks = extractTypeScriptBlocks(mdx)
    expect(blocks).toHaveLength(1)
    expect(blocks[0].code).toBe('const x = 1')
  })

  test('handles multi-line code blocks', () => {
    const mdx = [
      '```typescript',
      'import { CardinalSdk } from "@icure/cardinal-sdk"',
      '',
      'async function foo(sdk: CardinalSdk) {',
      '  return await sdk.patient.getPatient("123")',
      '}',
      '```',
    ].join('\n')

    const blocks = extractTypeScriptBlocks(mdx)
    expect(blocks).toHaveLength(1)
    expect(blocks[0].code).toContain('async function foo')
    expect(blocks[0].code).toContain('getPatient')
  })

  test('returns empty array for mdx with no typescript blocks', () => {
    const mdx = '# Title\n\nSome text\n\n```kotlin\nval x = 1\n```'
    expect(extractTypeScriptBlocks(mdx)).toHaveLength(0)
  })
})

describe('isImportLine', () => {
  test('detects standard import', () => {
    expect(isImportLine('import { Foo } from "bar"')).toBe(true)
  })

  test('detects type import', () => {
    expect(isImportLine('import type { Foo } from "bar"')).toBe(true)
  })

  test('rejects non-import lines', () => {
    expect(isImportLine('const x = 1')).toBe(false)
    expect(isImportLine('// import { Foo } from "bar"')).toBe(false)
  })
})

describe('stripImportsFromBlock', () => {
  test('removes import lines from block code', () => {
    const code = [
      'import { CardinalSdk } from "@icure/cardinal-sdk"',
      '',
      'const x = await sdk.patient.getPatient("123")',
    ].join('\n')

    const stripped = stripImportsFromBlock(code)
    expect(stripped).not.toContain('import')
    expect(stripped).toContain('const x = await sdk.patient.getPatient')
  })

  test('removes multi-line import statements', () => {
    const code = [
      'import {',
      '\tCardinalSdk,',
      '\tDecryptedPatient,',
      '} from "@icure/cardinal-sdk"',
      '',
      'const x = await sdk.patient.getPatient("123")',
    ].join('\n')

    const stripped = stripImportsFromBlock(code)
    expect(stripped).not.toContain('import')
    expect(stripped).not.toContain('CardinalSdk')
    expect(stripped).not.toContain('DecryptedPatient')
    expect(stripped).not.toContain('@icure/cardinal-sdk')
    expect(stripped).toContain('const x = await sdk.patient.getPatient')
  })

  test('preserves non-import lines exactly', () => {
    const code = [
      'const x = 1',
      'const y = 2',
    ].join('\n')

    expect(stripImportsFromBlock(code)).toBe(code)
  })
})
